feat: Database Monitoring PostgreSQL documentation [Feb Release] (#19932)
newrelic/docs-website: src/install/postgresql/linux/install-apt-query-monitoring.mdx

---
headingText: Enable query monitoring
componentType: default
---

> You can view query-level analysis to monitor your SQL queries within your PostgreSQL database to assess their performance and impact. This feature provides insights into execution times, resource consumption, and potential bottlenecks, enabling you to optimize database operations. For more information, see [Query-Level Analysis](#).

Reviewer: ```suggestion
You can view query-level analysis to assess the performance and impact of your SQL queries in your PostgreSQL database. This feature provides insights into execution times, resource consumption, and potential bottlenecks, enabling you to optimize database operations. For more information, refer [Query-Level Analysis](#).
```

**Prerequisites:**

> * Ensure that your PostgreSQL version is 12 or higher.

Reviewer: ```suggestion
* PostgreSQL version is 12 or higher.
```

    ```shell
    git clone https://github.com/postgrespro/pg_wait_sampling.git 
    cd pg_wait_sampling
    sudo make USE_PGXS=1
    sudo make USE_PGXS=1 install
    ```

3. To retrieve the individual query and execution plan details, install the `pg_stat_monitor` extension by running the following commands:

    ```shell
    git clone https://github.com/percona/pg_stat_monitor.git 
    cd pg_stat_monitor
    sudo make USE_PGXS=1
    sudo make USE_PGXS=1 install 
    ```

4. To retrieve slow running queries, install the `pg_stat_statements` extension:

    1. Open the `postgresql.conf` file:

        ```bash
        sudo nano /etc/postgresql/<your-version>/main/postgresql.conf
        ```

    2. Add `pg_stat_statements` to `shared_preload_libraries` in the `postgresql.conf` file:

        ```bash
        shared_preload_libraries = 'pg_stat_statements,pg_wait_sampling,pg_stat_monitor'
        pg_stat_statements.max = 10000    # Maximum number of statements tracked
        pg_stat_statements.track = all    # Track all statements
        pg_stat_statements.save = on      # Save statistics across server restarts
        ```

> 5. Restart the PostgreSQL server, run:

Reviewer: Should this be "To restart"?

    ```shell
    sudo systemctl restart postgresql
    ```

> 6. Connect to the PostgreSQL server, run:

Reviewer: ?To connect

    ```shell
    sudo -u postgres psql
    ```

> 7. Enable the extensions for your database, run:

Reviewer: ?To enable

    ```shell
    CREATE EXTENSION pg_stat_statements;
    CREATE EXTENSION pg_wait_sampling;
    CREATE EXTENSION pg_stat_monitor;
    ```

> 8. Grant the required permissions, run:

Reviewer: ?To grant
